### ISSUE TITLE
Adaptive Mechanism Design (AMD) algorithm pipeline introduction

### ISSUE THREAD

**Issue by quantaji** [quantaji/AMD-SSD#3]
# Adaptive Mechanism Design (AMD) algorithm pipeline

## theoretical derivation
From the paper, we know each agent's pseudo policy gradient assumed by the central planner
$$\nabla_{i, \text{pseudo}} J_i (\theta_i) = \sum_{\tau} \nabla_i \log \pi_{i}(\tau) \Psi_i(\tau),$$
where $\Psi_i(\tau)$ can be GAE estimator, or TD-error, the value function may also come from a critic. The actual parameter update is 
$$\nabla_{i, \text{actual}} J_i (\theta_i) = \sum_{\tau} \nabla_i \log \pi_{i}(\tau) \left[ \Psi_i(\tau) + r_{\text{planner}}(\tau, i)\right].$$
The reward given by planner is a function of current states, 
$$r_{\text{planner}}(\tau, i) \sim \pi_{\text{planner}}(s_1(\tau), \ldots, s_n(\tau); \theta_{p}) $$
The planner update its parameter by a similar policy gradient
$$\Delta \theta_p = \eta_p \sum_i \eta_i \nabla_{i, \text{pseudo}} J_i (\theta_i)\times \left(\sum_\tau \nabla_i \log \pi_i(\tau) \nabla_{p} \log\pi_{p}(\tau) r_{\text{planner}}(\tau, i) \right) $$
$$= \eta_p \sum_i \sum_\tau \eta_i \times \left( \nabla_{i, \text{pseudo}} J_i (\theta_i)\cdot \log \pi_i(\tau) \right) \times r_{\text{planner}}(\tau, i) \times \nabla_p \log\pi_{p}(\tau)$$
$$= \eta_p \sum_i \sum_\tau \eta_i a_i(\tau) \cdot r_{\text{planner}}(\tau, i) \cdot \nabla_p \log\pi_{p}(\tau).$$
Here I call $a_i(\tau) := \nabla_{i, \text{pseudo}} J_i (\theta_i)\cdot \log \pi_i(\tau)$ the **awarenss** of individual client, this is computed after sampling, but before passing to learning algorithm. One may also add the cost of central planner's reward by additional gradient
$$-\alpha \nabla_p \sum_i \sum_{\tau} \left\|r_{\text{planner}}(\tau, i)\right\|_2$$
## core implementation
- on algorithm init
    - Given a `pettingzoo.utils.env.ParallelEnv`, convert it to `ray.rllib.env.multi_agent_env.MultiAgentEnv`. [[source]](https://github.com/quantaji/FoRL-project/blob/31910217ee348a697ad377abad3134362b865cce/core/algorithms/amd/wrappers/pettingzoo_env_to_rllib_env.py#L9)
    - Generate a new environment with **central planner** as one of the agent. [[source1]](https://github.com/quantaji/FoRL-project/blob/31910217ee348a697ad377abad3134362b865cce/core/algorithms/amd/wrappers/rllib_env_with_central_planner.py#L15) [[source2]](https://github.com/quantaji/FoRL-project/blob/31910217ee348a697ad377abad3134362b865cce/core/algorithms/amd/amd.py#L128)
    - Initialize some constants (`gym.spaces.Space` for unflatten rewards), and identify central planner to the algorithm's worker. This is achieved by a rllib callback. [[source]](https://github.com/quantaji/FoRL-project/blob/31910217ee348a697ad377abad3134362b865cce/core/algorithms/amd/callback.py#L11)
- on sampling
    - after sampling, each individual agent post-process trajectories, compute GAE targets using critics. [[source]](https://github.com/quantaji/FoRL-project/blob/31910217ee348a697ad377abad3134362b865cce/core/algorithms/amd/amd_torch_policy.py#L101)
- on training step
    - each individual agent compute the awareness of this current batch. [[source]](https://github.com/quantaji/FoRL-project/blob/31910217ee348a697ad377abad3134362b865cce/core/algorithms/amd/amd_torch_policy.py#L159)
    - then individual agents and central planner communicate, exchange infomation on rewards from central planner, and awareness from agents. [[source]](https://github.com/quantaji/FoRL-project/blob/31910217ee348a697ad377abad3134362b865cce/core/algorithms/amd/amd.py#L252)
    - in the training step
        - agent update policy loss and critic's value loss. [[source]](https://github.com/quantaji/FoRL-project/blob/31910217ee348a697ad377abad3134362b865cce/core/algorithms/amd/amd_torch_policy.py#L122)
        - central planner compute the loss according to AMD algorithm, as shown in previous section. [[source]](https://github.com/quantaji/FoRL-project/blob/31910217ee348a697ad377abad3134362b865cce/core/algorithms/amd/amd_torch_policy.py#L214)

The whole training and configuration script is shown [here](https://github.com/quantaji/FoRL-project/blob/main/scripts/amd_wolfpack.py).

**Comment by quantaji** [comment]
In the original paper's implementation, the authors assume the most basic case of sigmoid parametrization ([source](https://github.com/longtermrisk/marltoolbox/blob/cae1ba94ccb44700b66a32e0734a0f11c9c6c7fe/marltoolbox/algos/adaptive_mechanism_design/planning_agent.py#L500)), $P(C)=\frac{\exp (\theta)}{1+\exp (\theta)}$, this simplifies a lot in calculating policy gradient for agents.

However, in our implementation, we give options to assume the agent's policy is "pretended" to be softmax parameterization $\pi(a|s) = \mathrm{softmax}(\theta_{s,a})$, and another option to exactly calculate the gradient ([source](https://github.com/quantaji/FoRL-project/blob/31910217ee348a697ad377abad3134362b865cce/core/algorithms/amd/amd_torch_policy.py#L166)).

I am very curious about these two option's difference.
